ktor-server/ktor-server-plugins/ktor-server-metrics/jvm/src/io/ktor/server/metrics/dropwizard/DropwizardMetrics.kt

```diff
@@ -14,129 +14,112 @@ import io.ktor.util.pipeline.*
-                    "jvm.attributes" to ::JvmAttributeGaugeSet
-                )
-                    .filter { (name, _) ->
-                        !configuration.registry.names.any { existingName -> existingName.startsWith(name) }
-                    }
-                    .forEach { (name, metric) -> configuration.registry.register(name, metric()) }
-            }
+    public var registerJvmMetricSets: Boolean = true
+}
 
-            val phase = PipelinePhase("DropwizardMetrics")
-            pipeline.insertPhaseBefore(ApplicationCallPipeline.Monitoring, phase)
-            pipeline.intercept(phase) {
-                plugin.before(call)
-                try {
-                    proceed()
-                } catch (e: Exception) {
-                    plugin.exception(call, e)
-                    throw e
-                } finally {
-                    plugin.after(call)
-                }
-            }
+private class RoutingMetrics(val name: String, val context: Timer.Context)
+private val routingMetricsKey = AttributeKey<RoutingMetrics>("metrics")
 
-            pipeline.environment.monitor.subscribe(Routing.RoutingCallStarted) { call ->
-                val name = call.route.toString()
-                val meter = plugin.registry.meter(name(plugin.baseName, name, "meter"))
-                val timer = plugin.registry.timer(name(plugin.baseName, name, "timer"))
-                meter.mark()
-                val context = timer.time()
-                call.attributes.put(
-                    routingMetricsKey,
-                    RoutingMetrics(name, context)
-                )
-            }
+private data class CallMeasure constructor(val timer: Timer.Context)
+private val measureKey = AttributeKey<CallMeasure>("metrics")
 
-            pipeline.environment.monitor.subscribe(Routing.RoutingCallFinished) { call ->
-                val routingMetrics = call.attributes.take(routingMetricsKey)
-                val status = call.response.status()?.value ?: 0
-                val statusMeter = plugin.registry.meter(name(plugin.baseName, routingMetrics.name, status.toString()))
-                statusMeter.mark()
-                routingMetrics.context.stop()
-            }
+private object BeforeMonitoring : Hook<(ApplicationCall) -> Unit> {
+    override fun install(application: ApplicationCallPipeline, handler: (ApplicationCall) -> Unit) {
+        val phase = PipelinePhase("MetricsBeforeMonitoring")
+        application.insertPhaseBefore(ApplicationCallPipeline.Monitoring, phase)
+        application.intercept(phase) {
+            handler(call)
+        }
+    }
+}
 
-            return plugin
+private object AfterMonitoring : Hook<(ApplicationCall) -> Unit> {
+    override fun install(application: ApplicationCallPipeline, handler: (ApplicationCall) -> Unit) {
+        application.intercept(ApplicationCallPipeline.Monitoring) {
+            try {
+                proceed()
```

Comment:
that's not after monitoring, but after whole pipeline is executed

Reply:
Am I correct that these 2 blocks are equivalent?
```kotlin
  application.intercept(ApplicationCallPipeline.Monitoring) {
      try {
          proceed()
      } finally {
          HANDLE_ACTION_FIRST()
      }
}
```
and
```kotlin
  application.intercept(ApplicationCallPipeline.Plugins) {
      try {
          proceed()
      } finally {
          HANDLE_ACTION_SECOND()
      }
}
```
The only difference is that if it breaks before `Plugins` phase then `HANDLE_ACTION_SECOND` will not be executed, which is not the case with `HANDLE_ACTION_FIRST` although it's a rare case.

Correct?

Reply:
Btw, not it seems that there is no actual difference between `on(CallFinished)`(which happens after the full pipeline execution if I'm not mistaken) and `on(AfterMonitoring)` (which is not "after monitoring" but after everything)

Reply:
Yes, it's correct